app/credentials/apikey/apikey.go

```diff
@@ -0,0 +1,140 @@
+package apikey
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"net/url"
+	"path"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc/credentials"
+)
+
+const (
+	IDHeader     = "tmprl-api-key-id"
+	SecretHeader = "tmprl-api-secret-key"
+	Separator    = "_"
+
+	RequestSignatureHeader           = "tmprl-request-signature"
+	RequestSignatureAlgorithmHeader  = "tmprl-request-signature-algorithm"
+	RequestDatetimeHeader            = "tmprl-request-datetime"
+	DefaultRequestSignatureAlgorithm = "tmprl-hmac-sha256"
+
+	// requestSignatureFormat requires each of the following data on a newline
+	// KeyID
+	// RequestSignatureAlgorithm
+	// RequestDatetime
+	// ActionName
+	requestSignatureFormat = "%s\n%s\n%s\n%s"
+)
+
+type Credential struct {
+	ID         string
+	secret     string // secret kept private to prevent accidental access.
+	enableHMAC bool
+	insecure   bool
+}
+
+type Option = func(c *Credential)
+
+func WithHMAC(enable bool) Option {
+	return func(c *Credential) {
+		c.enableHMAC = enable
+	}
+}
+
+func WithInsecure(insecure bool) Option {
+	return func(c *Credential) {
+		c.insecure = insecure
+	}
+}
+
+func NewCredential(key string, opts ...Option) (Credential, error) {
+	if len(key) == 0 {
+		return Credential{}, fmt.Errorf("an empty API key was provided")
+	}
+
+	s := strings.Split(key, Separator)
+	if len(s) < 2 {
+		return Credential{}, fmt.Errorf("an API key must be in the format of `{prefix}_{base62}_{base62]}`")
+	}
+
+	c := Credential{
+		ID:     s[len(s)-2],
+		secret: s[len(s)-1],
+	}
+	for _, opt := range opts {
+		opt(&c)
+	}
+
+	return c, nil
+}
+
+func (c Credential) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
+	ri, ok := credentials.RequestInfoFromContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("failed to retrieve request info from context")
+	}
+
+	if !c.insecure {
+		// Ensure the API key, AKA bearer token, is sent over a secure connection - meaning TLS.
+		if err := credentials.CheckSecurityLevel(ri.AuthInfo, credentials.PrivacyAndIntegrity); err != nil {
+			return nil, fmt.Errorf("the connection's transport security level is too low for API keys: %v", err)
+		}
+	}
+
+	if c.enableHMAC {
```

Comment:
this feels premature to check-in to the code base?

Reply:
It was previously in the codebase, so I wanted to move it along to the PerRPCCredentials so this PR wasn't destructive. Then my thought was we can remove it in a future PR when desired.